spec/abci++/abci++_properties_001_draft.md

```diff
@@ -0,0 +1,675 @@
+5. _p_'s Tendermint signs and broadcasts the Precommit message.
+
+In the cases when _p_'s Tendermint is to broadcast `precommit nil` messages (either _2f+1_ `prevote nil` messages received, or _timeoutPrevote_ triggered), _p_'s Tendermint does **not** call `RequestExtendVote` and will include an empty byte array as vote extension in the `precommit nil` message.
+
+### VerifyVoteExtension
+
+#### Parameters and Types
+
+>**TODO** (still controversial). Current status of discussions:
+
+* **Request**:
+
+    | Name      | Type  | Description                                                                              | Field Number |
+    |-----------|-------|------------------------------------------------------------------------------------------|--------------|
+    | extension | bytes | Sender Application's vote information to be validated                                    | 1            |
+    | hash      | bytes | The hash of the propsed block that the vote extension refers to                          | 2            |
+    | address   | bytes | [Address](../core/data_structures.md#address) of the validator that signed the extension | 3            |
+    | height    | int64 | Height of the block  (for sanity check).                                                 | 4            |
+
+* **Response**:
+
+    | Name   | Type | Description                                           | Field Number |
+    |--------|------|-------------------------------------------------------|--------------|
+    | accept | bool | If false, Application is rejecting the vote extension | 1            |
+
+* **Usage**:
+    * If `ResponseVerifyVoteExtension.accept` is false, Tendermint will flag the extension as _invalid_ but the
+      validity of the received Precommit message will not be influenced by this.
+    * The implementation of `VerifyVoteExtension` MUST be deterministic. Moreover, the value of
+      `VerifyVoteExtension.accept` MUST *exclusively* depend on the parameters passed in
+      the call to `RequestVerifyVoteExtension`, and the last committed Application state.
+
+#### When does Tendermint call it?
+
+When a validator _p_ is in Tendermint consensus round _r_, height _h_, state _prevote_ (**TODO** discuss: I think I must remove the state
+from this condition, but not sure), and _p_ receives a Precommit message for round _r_, height _h_ from _q_:
+
+1. _p_'s Tendermint calls `RequestVerifyVoteExtension`.
+2. The Application returns _accept_ or _reject_ via `ResponseVerifyVoteExtension.accept`.
+3. If the Application returns _reject_, _p_'s Tendermint flag the extension as _invalid_.
+4. _p_'s Tendermint keeps all received votes, together with their corresponding vote extensions
+   (and possibly _invalid_ flags) in its internal data structures. These will be used to:
+   * calculate field _LastCommitHash_ in the header of the block proposed for height _h + 1_
```

Comment:
Maybe a minor point, but only validators need to store these, right? Full nodes, which do not produce blocks, will just check that the `last_commit` included in the block matches the header data.